### ISSUE TITLE
Handle daemon restarts

### ISSUE THREAD

**Issue by odesenfans** [mhchia/py-libp2p-daemon-bindings#59]
Hello! Let's discuss a solution to support restarts of the P2P daemon. I'm currently experiencing memory leak issues with the JS P2P daemon which ultimately lead it to crash. While I hope this issue will be resolved in future versions, I need a fallback solution in the meantime.

The main issue is that the daemon "forgets" its configuration (subscribed topics, protocols, etc) on restarts. Code that uses the P2P client in my application times out and fails because the communication with the daemon must be reconfigured.

I came up with a solution, basically to create a `P2PClientFactory`/`P2PClientManager` class that stores a configured P2P client and provides it to the application tasks. If a typical exception occurs (ex: `IncompleteRead` when waiting for pubsub data), this factory class can then reinitialize a P2P client instance and provide it to the tasks. Example:

```python
async def publish_to_alive_topic_task(p2p_client_factory):
    topic = "my_app_alive_topic"
    data = b"hello, I'm a P2P app"

    while True:
        with p2p_client_factory.client() as p2p_client:
            await p2p_client.pubsub_publish(topic, data)

        await anyio.sleep(60)
        
async def main():
    async with P2PClientFactory(config) as p2p_client_factory:
        alive_task = asyncio.create_task(publish_to_alive_topic_task(p2p_client_factory)
```

The implementation of the class should look like this:

```python
class P2PClientFactory:
    def __init__(self, config):
        self.config = config
        self.p2p_client = None

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.p2p_client is not None:
            await self.p2p_client.close()

    @asynccontextmanager
    async def client(self):
        if self.p2p_client is None:
            self.p2p_client = P2PClient(self.config)
            await self.p2p_client.start_listening()
        try:
            yield self.p2p_client
        except IncompleteRead:
            await self.p2p_client.close()
            self.p2p_client = None
```

The main problem I have with this design at the moment is that it needs a few modifications in `py-libp2p-daemon-bindings`:

1. At the moment, the P2P client can only be initialized properly within an async generator:

```python
p2p_client = P2PClient(config)
async with p2p_client.listen():
  # Do stuff with p2p_client
```

As stated above, I'd need functions to control the lifetime of the P2P client manually to avoid closing the client as long as the daemon is working fine.

2. To achieve this objective, the task group used in the `ControlClient` gets in the way. I'm not aware of any simple solution in anyio to create a task group without a context manager. Any idea?

So, basically: what do you think of the concept, and do you have proposals to improve on it?

**Comment by mhchia** [comment]
Sorry, it was a busy week for me. I will get back to you tomorrow.

**Comment by mhchia** [comment]
Sorry for being late again.

> As stated above, I'd need functions to control the lifetime of the P2P client manually to avoid closing the client as long as the daemon is working fine.

Will it work for you if `P2PClientFactory.client` looks like this? Let me know if I didn't understand it clearly and I can explore other approaches.
```python

    @asynccontextmanager
    async def client(self):
        while True:
            if self.p2p_client is None:
                # Do something with config and p2p client.
                self.p2p_client = P2PClient(self.config)
            async with self.p2p_client.listen():
                try:
                    yield self.p2p_client
                except IncompleteRead:
                    await self.p2p_client.close()
                    self.p2p_client = None
                except Exception:
                    # Don't catch other exceptions.
                    raise
                else:
                    # Nothing is raised and yield is returned. Seems the listener is closed?
                    # Ref: https://github.com/mhchia/py-libp2p-daemon-bindings/blob/299b3dbf1b95844a49c4ff42b1694b43fa4f93e0/p2pclient/control.py#L120
                    break
```

> 2. To achieve this objective, the task group used in the ControlClient gets in the way. I'm not aware of any simple solution in anyio to create a task group without a context manager. Any idea?

I'm not aware of it either. I guess that's the [preferred way](https://anyio.readthedocs.io/en/stable/tasks.html#creating-and-managing-tasks) of anyio to spawn new tasks. Though we should be able to avoid the `async with` statement which is required to create a task group by calling `__aenter__` on the task group manually and calling `__aexit__` somewhere else.

**Comment by odesenfans** [comment]
Hello, no worries :) The proposed solution does not do the trick because I'd like this factory to support providing the same client for multiple tasks at the same time. In my example, I can have several tasks publishing on the same topic. I want the client listener tasks to be spawned exactly once, but the client can be used from several tasks at the same time to send messages.

Alternatively, I'd need a solution to spawn multiple `P2PClient` instances communicating with the same daemon.

**Comment by odesenfans** [comment]
To add on this topic, this idea can basically be split in two parts:

### Support daemon restarts

* Factory class
* Make it possible to restart the listener tasks

### Thread-safe client

* Add a mutex to the client class to ensure that writes to the daemon are atomic, as they use the same socket.

**Comment by mhchia** [comment]
Do you need to register a stream handler? It looks like you need a single control client for all tasks under this condition. Otherwise, since other methods always create a new connection to the daemon, it should be safe for different tasks to share a `ControlClient`.

**Comment by mhchia** [comment]
Another issue for `P2PClientFactory.client` to return a single p2pclient is, if p2pclient fails and is re-created, other tasks might still reference the old client.

I'm thinking we can still have a single method `P2PClientFactory.listen` which maintains the p2pclient for us (if p2plcient fails, just create it again). And we don't access the client with `P2PClientFactory.listen`. Instead, one needs to acquire the async lock inside `P2PClientFactory` and access the client with `P2PClientFactory.p2pclient`. Will it work for you?

**Comment by mhchia** [comment]
No matter what, the points in https://github.com/mhchia/py-libp2p-daemon-bindings/issues/59#issuecomment-1108675581 all sound good to me!:)

**Comment by odesenfans** [comment]
So, I had a chat with a maintainer of the JS daemon the other day at Paris P2P. Turns out that they're mostly using the daemon for testing use cases at the moment. So I'm currently in the process of writing bindings to the Rust implementation of libp2p. I'll look into this issue only if I don't manage to make the bindings work.

**Comment by mhchia** [comment]
@odesenfans I see. It makes great sense to use the binding to rust-libp2p directly if the communication between rust and python work fine. Sorry for the late responses due to my lack of bandwidth. You really did a good job on improving the codebase. Please let me know in the future if you still would like to extend or take over this project in the future.

**Comment by mhchia** [comment]
@odesenfans I'm cleaning the permission recently since py-libp2p folks might take over the project, per the discussion in https://github.com/libp2p/py-libp2p/pull/973. Thanks for all your valuable contributions!

**Comment by mhchia** [comment]
Closed since it's stale I believe

**Comment by mhchia** [comment]
whoops I mistook this issue with another PR. Open it again
